### PR TITLE
adjust stagingBufferPool size when recieve memory warning #3733

### PR DESCRIPTION
**Pull request by hana-alice**
-add listener for memory warning and resize staging buffer pool;
-remove reset() in acquire but reset by the end of MTLCommandBuffer;

tested both on IOS14.2(iphone SE2) and MAC10.15.7, performance test case

### CODE REVIEW

File: cocos/renderer/gfx-metal/MTLCommandBuffer.mm [hunk 1]
```diff
@@ -303,14 +303,18 @@
...
     memcpy(stagingBuffer.mappedData, data, size);
     id<MTLBlitCommandEncoder> encoder = [_mtlCommandBuffer blitCommandEncoder];
     [encoder copyFromBuffer:stagingBuffer.mtlBuffer
                sourceOffset:stagingBuffer.startOffset
                    toBuffer:((CCMTLBuffer *)buff)->getMTLBuffer()
           destinationOffset:offset
                        size:size];
+    [_mtlCommandBuffer addCompletedHandler:^(id<MTLCommandBuffer> buf){
+        bufferPool->reset();
+    }];
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLCommandBuffer.mm:317]: already handle the complete event in `MTLDevice::preset()`.

File: cocos/renderer/gfx-metal/MTLDevice.h [hunk 2]
```diff
@@ -48,6 +48,9 @@ class CCMTLDevice : public Device {
     CC_INLINE CCMTLGPUStagingBufferPool *gpuStagingBufferPool() const { return _gpuStagingBufferPools[_currentFrameIndex]; }
     CC_INLINE bool isSamplerDescriptorCompareFunctionSupported() const { return _isSamplerDescriptorCompareFunctionSupported; }
 
+private:
+    void responseToMemoryAlarm();
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLDevice.h:52]: `onMemoryWarning` is a better name?

File: cocos/renderer/gfx-metal/MTLDevice.h [hunk 3]
```diff
@@ -62,6 +65,7 @@ class CCMTLDevice : public Device {
     CCMTLGPUStagingBufferPool *_gpuStagingBufferPools[MAX_FRAMES_IN_FLIGHT] = { nullptr };
     CCMTLSemaphore *_inFlightSemaphore = nullptr;
     uint _currentFrameIndex = 0;
+    uint32_t    _memoryAlarmListenerId = 0;
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLDevice.h:68]: unneeded spaces.

File: cocos/renderer/gfx-metal/MTLDevice.mm [hunk 4]
```diff
@@ -347,5 +354,11 @@
     static_cast<CCMTLCommandBuffer *>(_cmdBuff)->copyBuffersToTexture(buffers, texture, regions, count);
 }
 
+void CCMTLDevice::onMemoryWarning() {
+    for (int i = 0; i < MAX_FRAMES_IN_FLIGHT; ++i) {
+        _gpuStagingBufferPools[i]->shrinkSize();
+    }
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLDevice.mm:360]: How do you know which staging buffer pool can be shrink?

Comment (hana-alice) [reply cocos/renderer/gfx-metal/MTLDevice.mm:360]: it's kind of "try-to-shrink", if all the buffers in the buffer pool are in use, shrinkSize does nothing. Only when pool size greater than 1 and there are buffers can be released(curOffet == 0), shrinkSize release and erase them from buffer pool.

Comment (minggo) [reply cocos/renderer/gfx-metal/MTLDevice.mm:360]: Sounds good.

**Pull request merged**